### PR TITLE
WT-5769 Search history store can potentially walk the whole history store tree

### PR DESCRIPTION
**Pull request by tetsuo-cpp**
Here's a rough draft of the change we discussed. I'm sure I've missed an edge case or ten but I think this is enough to show that it's a viable approach.

### CODE REVIEW

**Comment by tetsuo-cpp** [comment]
When I run without this change:
```
alexc@alexc:~/work/wiredtiger/build_posix$ LD_LIBRARY_PATH=.libs python ../test/suite/run.py hs_14
FERROR in test_hs_latency.test_hs14.test_hs14

======================================================================
FAIL: test_hs_latency.test_hs14.test_hs14
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/data0/alexc/work/wiredtiger/test/suite/test_hs_latency.py", line 100, in test_hs14
    visible_hs_latency, invisible_hs_latency
AssertionError: Reader took an order of magnitude longer for when all history store records were invisible, visible=0.170781850815, invisible=176.184886932

----------------------------------------------------------------------
Ran 1 test in 178.261s

FAILED (failures=1)
```

**Comment by tetsuo-cpp** [comment]
> This change looks pretty good to me, and seems like a nice simplification overall (which is a bit of a surprise). I did think that there would also be a change in __wt_hs_cursor_position to avoid setting the read_timestamp - or is that orthogonal?

I believe that part is just choosing what timestamp to use in the `search_near` as part of the key. Even with this new scheme, you'd still want to do this.

File: src/include/txn.i [hunk 1]
```diff
@@ -737,7 +737,7 @@ __wt_txn_upd_visible_type(WT_SESSION_IMPL *session, WT_UPDATE *upd)
         if (prepare_state == WT_PREPARE_LOCKED)
             continue;
 
-        if (F_ISSET(session, WT_SESSION_HS_IGNORE_VISIBILITY) && upd->txnid != WT_TXN_ABORTED &&
+        if (WT_IS_HS(S2BT(session)) && upd->txnid != WT_TXN_ABORTED &&
           upd->type == WT_UPDATE_STANDARD) {
             /* If we are resolving a modify then the btree must be the history store. */
```
Comment (quchenhao) [inline src/include/txn.i:742]: We should change the comment.

File: src/include/txn.i [hunk 2]
```diff
@@ -975,13 +975,7 @@ __wt_txn_read(WT_SESSION_IMPL *session, WT_CURSOR_BTREE *cbt, WT_ITEM *key, uint
     }
 
     /* If the start time point is visible then we need to return the ondisk value. */
-    if (F_ISSET(session, WT_SESSION_HS_IGNORE_VISIBILITY) ||
```
Comment (kommiharibabu) [inline src/include/txn.i:978]: Can we simplify the condition at lines 863 and 955 with this change?

File: src/include/session.h [hunk 3]
```diff
@@ -166,33 +166,32 @@ struct __wt_session_impl {
...
+#define WT_SESSION_LOCKED_TURTLE 0x00080000u
+#define WT_SESSION_LOGGING_INMEM 0x00100000u
+#define WT_SESSION_NO_DATA_HANDLES 0x00200000u
+#define WT_SESSION_NO_LOGGING 0x00400000u
+#define WT_SESSION_NO_RECONCILE 0x00800000u
+#define WT_SESSION_NO_SCHEMA_LOCK 0x01000000u
+#define WT_SESSION_QUIET_CORRUPT_FILE 0x02000000u
+#define WT_SESSION_READ_WONT_NEED 0x04000000u
+#define WT_SESSION_RESOLVING_TXN 0x08000000u
+#define WT_SESSION_ROLLBACK_TO_STABLE 0x10000000u
```
Comment (kommiharibabu) [inline src/include/session.h:192]: With this change in place, do we still need the `rollback to stable` flag that was used for skipping the tombstones? Without this change itself, the cursor ignore flag should enough to remove this session flag I feel.

**Comment by agorrod** [comment]
```
I believe that part is just choosing what timestamp to use in the search_near as part of the key. Even with this new scheme, you'd still want to do this.
```

I see - you are right.

File: src/include/txn.i (hunk 1 above)
Comment (tetsuo-cpp) [reply src/include/txn.i]: df40364

File: src/include/txn.i (hunk 2 above)
Comment (tetsuo-cpp) [reply src/include/txn.i:978]: 91e8023

File: src/include/session.h (hunk 3 above)
Comment (tetsuo-cpp) [reply src/include/session.h:192]: As discussed, we still need this flag to figure out what error code to return when we fail to read a data file (yeah, it's gross). We're planning to address this in WT-5832.

**Comment by tetsuo-cpp** [comment]
I promoted this PR from a draft just before.

I was mostly concerned about rollback to stable and the prepare fixup process. But after talking to @kommiharibabu, I don't think we move visibility checks there (in fact, we explicitly want to avoid that since that's a root cause of another bug we had).

**Pull request merged**